keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,524 @@
+
+For the reference KMS plugin, the encrypted local KEK is stored in etcd via the `metadata` field, and once decrypted, it can be stored in memory as part of the KMS plugin cache to be used for encryption and decryption of DEKs. The encrypted local KEK is used as the key and the decrypted local KEK is stored as the value.
+
+```proto
+message EncryptResponse {
+    // The encrypted data.
+    bytes ciphertext = 1;
+    // The KMS key ID used for encryption operations.
+    // This can be used to drive rotation.
+    string current_key_id = 2;
+    // Additional metadata to be stored with the encrypted data.
+    // This metadata can contain the encrypted local KEK that was used to encrypt the DEK.
+    // Stored unencrypted in etcd.
+    map<string, string> metadata = 3;
+}
+```
+
+The `DecryptRequest` passes the same `current_key_id` and `metadata` returned by the previous `EncryptResponse` of this data as its `observed_key_id` and `metadata` for the decryption request.
+
+```proto
+message DecryptRequest {
+    // The data to be decrypted.
+    bytes ciphertext = 2;
+    // UID is a unique identifier for the request.
+    string uid = 3;
+    // The keyID that was provided to the apiserver during encryption.
+    // This represents the KMS KEK that was used to encrypt the data.
+    string observed_key_id = 4;
+    // Additional metadata that was sent by the KMS plugin during encryption.
+    map<string, string> metadata = 5;
+}
+
+message DecryptResponse {
+    // The decrypted data.
+    bytes plaintext = 1;
+    // The KMS key ID used to decrypt the data.
+    string current_key_id = 2;
+    // Additional metadata that was sent by the KMS plugin.
+    map<string, string> metadata = 3;
+}
+
+message EncryptRequest {
+    // The data to be encrypted.
+    bytes plaintext = 2;
+    // UID is a unique identifier for the request.
+    string uid = 3;
+}
+```
+
+In terms of storage, a new structured protobuf format is proposed. Similar to the proto serializer, it will use a magic number to detect when the stored data is in a format that it understands:
+
+```go
+encryptedProtoEncodingPrefix = []byte{'e', 'k', '8', 's', 0}
+```
+
+The last byte represents the encoding style, with 0 meaning that the rest of the byte stream is a proto message of type `EncryptedObject`:
+
+```go
+type EncryptedObject struct {
+    TypeMeta `json:",inline" protobuf:"bytes,1,opt,name=typeMeta"`
+
+    EncryptResponse EncryptResponse `protobuf:"bytes,2,opt,name=encryptResponse"`
```

Comment:
I don't like that we are embedding EncryptResponse in EncryptedObject. I'd prefer if we just copy the fields we need (maybe wrapped in a different message) so that the storage format and the KMS plugin API can evolve independently.